src/userobjects/MoabSkinner.C

```diff
@@ -0,0 +1,1008 @@
+  skinner->find_skin(0, region, false, tris, &rtris);
+
+  // Create surface sets for the forwards tris
+  VolData vdata = {volume_set, Sense::FORWARDS};
+  createSurfaces(tris, vdata, surf_id);
+
+  // Create surface sets for the reversed tris
+  vdata.sense = Sense::BACKWARDS;
+  createSurfaces(rtris, vdata, surf_id);
+}
+
+void
+MoabSkinner::createSurfaces(moab::Range & faces, VolData & voldata, unsigned int & surf_id)
+{
+  if (faces.empty())
+    return;
+
+  // Loop over the surfaces we have already created
+  for (const auto & surfpair : surfsToVols)
+  {
+    // Local copies of surf/vols
+    moab::EntityHandle surf = surfpair.first;
+    std::vector<VolData> vols = surfpair.second;
+
+    // First get the entities in this surface
+    moab::Range tris;
+    _moab->get_entities_by_handle(surf, tris);
+
+    // Find any tris that live in both surfs
+    moab::Range overlap = moab::intersect(tris, faces);
+    if (!overlap.empty())
+    {
+      // Check if the tris are a subset or the entire surf
+      if (tris.size() == overlap.size())
+      {
+        // Whole surface -> Just need to update the volume relationships
+        updateSurfData(surf, voldata);
+      }
+      else
+      {
+        // If overlap is subset, subtract shared tris from this surface and create a new shared
+        // surface
+        _moab->remove_entities(surf, overlap);
+
+        // Append our new volume to the list that share this surf
+        vols.push_back(voldata);
+
+        // Create a new shared surface
+        moab::EntityHandle shared_surf;
+        surf_id++;
+        createSurf(surf_id, shared_surf, overlap, vols);
+      }
+
+      // Subtract from the input list
+      for (auto & shared : overlap)
+        faces.erase(shared);
+
+      if (faces.empty())
+        break;
+    }
+  }
+
+  if (!faces.empty())
+  {
+    moab::EntityHandle surface_set;
+    std::vector<VolData> voldatavec(1, voldata);
+    surf_id++;
+    createSurf(surf_id, surface_set, faces, voldatavec);
+  }
+}
+
+void
+MoabSkinner::buildGraveyard(unsigned int & vol_id, unsigned int & surf_id)
+{
+  // Create the graveyard group
+  moab::EntityHandle graveyard;
+  unsigned int id = nBins() + 1;
+  createGroup(id, "mat:Graveyard", graveyard);
+
+  // Create a volume set
+  moab::EntityHandle volume_set;
+  createVol(++vol_id, volume_set, graveyard);
+
+  // Set up for the volume data to pass to surfs
+  VolData vdata = {volume_set, Sense::FORWARDS};
+
+  // Find a bounding box
+  BoundingBox bbox = MeshTools::create_bounding_box(mesh());
+
+  // Build the two cubic surfaces defining the graveyard
+  createSurfaceFromBox(bbox, vdata, surf_id, false /* normals point into box */, _graveyard_scale_inner);
+  createSurfaceFromBox(bbox, vdata, surf_id, true /* normals point out of box */, _graveyard_scale_outer);
+}
+
+void
+MoabSkinner::createSurfaceFromBox(const BoundingBox & box,
+                                     VolData & voldata,
```

Comment:
Pretty sure this one can be `const` as well.

Reply:
Yep!